Package Request:dgsh(Directed Graph Shell)

<!--
IMPORTANT: we are not accepting issues about Termux
installations running on Android 5.x-6.x.
-->

**Package description**
<!--
Describe what this package do and why it should be
available. Make sure that requested package supports
cross-compilation and can run on AArch64 and ARM.
-->
dgsh is an advanced shell for programming higly complex pipe constructs for data flow in shell scripting. And hence it is the most essential tool for programmers, scripters, and system administrators. Having it on termux is highly needed.

**Link to home page and sources**
1. Home page:https://www.spinellis.gr/sw/dgsh/
2. Source code:https://github.com/dspinellis/dgsh

**Additional information**
Have you compiled or tried to compile the package on device?
Developer have not made or compiled this shell for arm and aarch64 architecture.
Added some of commits with official developer. But some errors cant be resolved. Help needed.
https://github.com/dspinellis/dgsh/issues/113

However, source files needed to be patched for arm/aarch64 so that it can be used on termux.

Having dgsh in termux is the most prominent requirement right now and it will prove to be highly beneficial for complex shell scripting purposes.

* If it didn't work then please provide the error you ran into.
* If it did work then please share any patches that you had to apply.

```
./lib/stdio-impl.h:66:22: error: field has incomplete type 'struct __sbuf'
lib/freading.c:40:14: error: incomplete definition of type 'struct __sFILE'
lib/freading.c:40:25: error: use of undeclared identifier '__SRD'
3 errors generated.
```

Help solve this error: 
Below may help
https://lists.gnu.org/archive/html/bug-gnulib/2018-05/msg00103.html

```
make -C core-tools CFLAGS=""
make[1]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
make  all-recursive
make[2]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
Making all in src
make[3]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/src'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/src'
Making all in .
make[3]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
make[3]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
Making all in tests
make[3]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/tests'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/tests'
make[2]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
make[1]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools'
cd core-tools/src && make build-install
make[1]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/src'
mkdir -p ../../build/bin ../../build/libexec/dgsh
cp dgsh-monitor dgsh-httpval dgsh-readval dgsh-merge-sum ../../build/bin/
cp dgsh-tee dgsh-writeval dgsh-readval dgsh-monitor dgsh-conc dgsh-wrap dgsh-enumerate dgsh-pecho dgsh-fft-input dgsh-w dgsh-parallel perm ../../build/libexec/dgsh/
make[1]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/core-tools/src'
make -C unix-tools make MAKEFLAGS=
make[1]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools'
make -C bash CFLAGS=""
make[2]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/bash'

	  ***********************************************************
	  *                                                         *
	  * GNU bash, version 4.3.46(1)-release (aarch64-unknown-linux-gnu)
	  *                                                         *
	  ***********************************************************

make[2]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/bash'
make -C coreutils
make[2]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
make  all-recursive
make[3]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
Making all in po
make[4]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils/po'
make[4]: Nothing to be done for 'all'.
make[4]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils/po'
Making all in .
make[4]: Entering directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
  CC       lib/freading.o
  CC       lib/freadseek.o
  CC       lib/freopen-safer.o
  CC       lib/ftoastr.o
  CC       lib/full-read.o
  CC       lib/full-write.o
  CC       lib/gethrxtime.o
  CC       lib/xtime.o
  CC       lib/getndelim2.o
lib/getndelim2.c:164:15: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
          if (GETNDELIM2_MAXIMUM < newsize - offset)
              ^~~~~~~~~~~~~~~~~~
lib/getndelim2.c:64:29: note: expanded from macro 'GETNDELIM2_MAXIMUM'
#define GETNDELIM2_MAXIMUM (PTRDIFF_MAX < SSIZE_MAX ? PTRDIFF_MAX : SSIZE_MAX)
                            ^~~~~~~~~~~
./lib/stdint.h:520:5: note: expanded from macro 'PTRDIFF_MAX'
    _STDINT_MAX (1, 64, 0l)
    ^~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:126:8: note: expanded from macro '_STDINT_MAX'
   ? ~ _STDINT_MIN (signed, bits, zero) \
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:122:31: note: expanded from macro '_STDINT_MIN'
  ((signed) ? (- ((zero) + 1) << ((bits) ? (bits) - 1 : 0)) : (zero))
               ~~~~~~~~~~~~~~ ^
lib/getndelim2.c:166:44: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
              size_t newsizemax = offset + GETNDELIM2_MAXIMUM + 1;
                                           ^~~~~~~~~~~~~~~~~~
lib/getndelim2.c:64:29: note: expanded from macro 'GETNDELIM2_MAXIMUM'
#define GETNDELIM2_MAXIMUM (PTRDIFF_MAX < SSIZE_MAX ? PTRDIFF_MAX : SSIZE_MAX)
                            ^~~~~~~~~~~
./lib/stdint.h:520:5: note: expanded from macro 'PTRDIFF_MAX'
    _STDINT_MAX (1, 64, 0l)
    ^~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:126:8: note: expanded from macro '_STDINT_MAX'
   ? ~ _STDINT_MIN (signed, bits, zero) \
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:122:31: note: expanded from macro '_STDINT_MIN'
  ((signed) ? (- ((zero) + 1) << ((bits) ? (bits) - 1 : 0)) : (zero))
               ~~~~~~~~~~~~~~ ^
2 warnings generated.
  CC       lib/gettime.o
  CC       lib/getugroups.o
  CC       lib/hard-locale.o
  CC       lib/hash.o
lib/hash.c:549:11: warning: implicit conversion from 'unsigned long' to 'float' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]
      if (SIZE_MAX <= new_candidate)
          ^~~~~~~~ ~~
./lib/stdint.h:543:20: note: expanded from macro 'SIZE_MAX'
# define SIZE_MAX  _STDINT_MAX (0, 64, 0ul)
                   ^~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:130:71: note: expanded from macro '_STDINT_MAX'
     ((((zero) + 1) << ((bits) ? (bits) - 1 - (signed) : 0)) - 1) * 2 + 1)
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
lib/hash.c:1079:15: warning: implicit conversion from 'unsigned long' to 'float' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]
          if (SIZE_MAX <= candidate)
              ^~~~~~~~ ~~
./lib/stdint.h:543:20: note: expanded from macro 'SIZE_MAX'
# define SIZE_MAX  _STDINT_MAX (0, 64, 0ul)
                   ^~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdint.h:130:71: note: expanded from macro '_STDINT_MAX'
     ((((zero) + 1) << ((bits) ? (bits) - 1 - (signed) : 0)) - 1) * 2 + 1)
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
2 warnings generated.
  CC       lib/hash-pjw.o
  CC       lib/hash-triple.o
  CC       lib/heap.o
  CC       lib/human.o
  CC       lib/i-ring.o
  CC       lib/idcache.o
  CC       lib/ino-map.o
  CC       lib/imaxtostr.o
  CC       lib/inttostr.o
  CC       lib/offtostr.o
  CC       lib/uinttostr.o
  CC       lib/umaxtostr.o
  CC       lib/ldtoastr.o
  CC       lib/linebuffer.o
  CC       lib/localcharset.o
  CC       lib/glthread/lock.o
In file included from lib/glthread/lock.c:23:
./lib/glthread/lock.h:147:17: warning: weak identifier 'pthread_cancel' never declared
#   pragma weak pthread_cancel
                ^
1 warning generated.
  CC       lib/long-options.o
  CC       lib/malloca.o
  CC       lib/math.o
  CC       lib/mbchar.o
  CC       lib/mbiter.o
  CC       lib/mbsalign.o
  CC       lib/mbscasecmp.o
  CC       lib/mbslen.o
  CC       lib/mbsstr.o
  CC       lib/mbswidth.o
  CC       lib/mbuiter.o
  CC       lib/memcasecmp.o
  CC       lib/memchr2.o
  CC       lib/memcmp2.o
  CC       lib/memcoll.o
  CC       lib/mgetgroups.o
  CC       lib/mkancesdirs.o
  CC       lib/dirchownmod.o
  CC       lib/mkdir-p.o
  CC       lib/modechange.o
  CC       lib/mpsort.o
  CC       lib/nproc.o
  CC       lib/openat-die.o
  CC       lib/openat-safer.o
  CC       lib/parse-datetime.o
./lib/parse-datetime.y:876:17: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
  if (24 * 60 < abs (n_minutes))
                ^
./lib/parse-datetime.y:876:17: note: use function 'labs' instead
  if (24 * 60 < abs (n_minutes))
                ^~~
                labs
1 warning generated.
  CC       lib/physmem.o
  CC       lib/posixtm.o
  CC       lib/posixver.o
  CC       lib/printf-frexp.o
  CC       lib/printf-frexpl.o
  CC       lib/priv-set.o
  CC       lib/progname.o
  CC       lib/propername.o
  CC       lib/qcopy-acl.o
  CC       lib/qset-acl.o
  CC       lib/quotearg.o
  CC       lib/randint.o
  CC       lib/randperm.o
  CC       lib/randread.o
  CC       lib/rand-isaac.o
  CC       lib/read-file.o
  CC       lib/readtokens.o
  CC       lib/readtokens0.o
  CC       lib/root-dev-ino.o
  CC       lib/safe-read.o
  CC       lib/safe-write.o
  CC       lib/same.o
  CC       lib/save-cwd.o
  CC       lib/savedir.o
  CC       lib/savewd.o
  CC       lib/se-context.o
  CC       lib/se-selinux.o
  CC       lib/settime.o
  CC       lib/sig-handler.o
  CC       lib/sockets.o
  CC       lib/stat-time.o
  CC       lib/statat.o
  CC       lib/mkstemp-safer.o
  CC       lib/strftime.o
  CC       lib/striconv.o
  CC       lib/strnlen1.o
  CC       lib/strintcmp.o
  CC       lib/strnumcmp.o
  CC       lib/sys_socket.o
  CC       lib/tempname.o
  CC       lib/glthread/threadlib.o
  CC       lib/timespec.o
  CC       lib/glthread/tls.o
In file included from lib/glthread/tls.c:21:
./lib/glthread/tls.h:79:17: warning: weak identifier 'pthread_cancel' never declared
#   pragma weak pthread_cancel
                ^
1 warning generated.
  CC       lib/trim.o
  CC       lib/u64.o
  CC       lib/unicodeio.o
  CC       lib/unistd.o
  CC       lib/dup-safer.o
  CC       lib/fd-safer.o
  CC       lib/pipe-safer.o
  CC       lib/unistr/u8-mbtoucr.o
  CC       lib/unistr/u8-uctomb.o
  CC       lib/unistr/u8-uctomb-aux.o
  CC       lib/uniwidth/width.o
  CC       lib/userspec.o
  CC       lib/utimecmp.o
lib/utimecmp.c:292:36: warning: bitwise negation of a boolean expression; did you mean logical negation? [-Wbool-operation]
                time_t s = src_s & ~ (res == 2 * BILLION);
                                   ^~~~~~~~~~~~~~~~~~~~~~
                                   !
lib/utimecmp.c:371:16: warning: bitwise negation of a boolean expression; did you mean logical negation? [-Wbool-operation]
      src_s &= ~ (res == 2 * BILLION);
               ^~~~~~~~~~~~~~~~~~~~~~
               !
2 warnings generated.
  CC       lib/utimens.o
  CC       lib/verror.o
  CC       lib/version-etc.o
  CC       lib/version-etc-fsf.o
  CC       lib/wctype-h.o
  CC       lib/write-any-file.o
  CC       lib/xmalloc.o
  CC       lib/xalloc-die.o
  CC       lib/xdectoimax.o
  CC       lib/xdectoumax.o
  CC       lib/xfreopen.o
  CC       lib/xfts.o
  CC       lib/xgetcwd.o
  CC       lib/xgetgroups.o
  CC       lib/xgethostname.o
  CC       lib/xmemcoll.o
  CC       lib/xnanosleep.o
  CC       lib/xprintf.o
  CC       lib/xreadlink.o
  CC       lib/xsize.o
  CC       lib/xstriconv.o
  CC       lib/xstrndup.o
  CC       lib/xstrtod.o
  CC       lib/xstrtoimax.o
  CC       lib/xstrtol.o
  CC       lib/xstrtoul.o
  CC       lib/xstrtol-error.o
  CC       lib/xstrtold.o
  CC       lib/xstrtoumax.o
  CC       lib/xvasprintf.o
  CC       lib/xasprintf.o
  CC       lib/yesno.o
  CC       lib/asnprintf.o
  CC       lib/asprintf.o
  CC       lib/at-func2.o
  CC       lib/calloc.o
  CC       lib/chdir-long.o
  CC       lib/euidaccess.o
  CC       lib/fclose.o
  CC       lib/fcntl.o
  CC       lib/fflush.o
  CC       lib/fnmatch.o
  CC       lib/freadahead.o
  CC       lib/freadptr.o
  CC       lib/frexp.o
  CC       lib/frexpl.o
  CC       lib/fseek.o
  CC       lib/fseeko.o
  CC       lib/fseterr.o
  CC       lib/fsusage.o
  CC       lib/fts.o
  CC       lib/futimens.o
  CC       lib/getcwd.o
  CC       lib/getcwd-lgpl.o
  CC       lib/getdtablesize.o
  CC       lib/getloadavg.o
  CC       lib/getpass.o
  CC       lib/getusershell.o
  CC       lib/group-member.o
  CC       lib/isapipe.o
  CC       lib/isnanf.o
  CC       lib/isnanl.o
  CC       lib/linkat.o
  CC       lib/mbrlen.o
  CC       lib/mbrtowc.o
  CC       lib/mktime.o
  CC       lib/mountlist.o
lib/mountlist.c:534:26: warning: implicit declaration of function 'makedev' is invalid in C99 [-Wimplicit-function-declaration]
            me->me_dev = makedev (devmaj, devmin);
                         ^
1 warning generated.
  CC       lib/nanosleep.o
  CC       lib/nl_langinfo.o
lib/nl_langinfo.c:151:1: error: redefinition of 'nl_langinfo'
nl_langinfo (nl_item item)
^
/data/data/com.termux/files/usr/include/langinfo.h:177:14: note: previous definition is here
static char *nl_langinfo(nl_item item)
             ^
lib/nl_langinfo.c:175:10: error: use of undeclared identifier 'GROUPING'
    case GROUPING:
         ^
lib/nl_langinfo.c:283:10: error: use of undeclared identifier 'INT_CURR_SYMBOL'
    case INT_CURR_SYMBOL:
         ^
lib/nl_langinfo.c:285:10: error: use of undeclared identifier 'MON_DECIMAL_POINT'
    case MON_DECIMAL_POINT:
         ^
lib/nl_langinfo.c:287:10: error: use of undeclared identifier 'MON_THOUSANDS_SEP'
    case MON_THOUSANDS_SEP:
         ^
lib/nl_langinfo.c:289:10: error: use of undeclared identifier 'MON_GROUPING'
    case MON_GROUPING:
         ^
lib/nl_langinfo.c:291:10: error: use of undeclared identifier 'POSITIVE_SIGN'
    case POSITIVE_SIGN:
         ^
lib/nl_langinfo.c:293:10: error: use of undeclared identifier 'NEGATIVE_SIGN'
    case NEGATIVE_SIGN:
         ^
lib/nl_langinfo.c:295:10: error: use of undeclared identifier 'FRAC_DIGITS'
    case FRAC_DIGITS:
         ^
lib/nl_langinfo.c:297:10: error: use of undeclared identifier 'INT_FRAC_DIGITS'
    case INT_FRAC_DIGITS:
         ^
lib/nl_langinfo.c:299:10: error: use of undeclared identifier 'P_CS_PRECEDES'
    case P_CS_PRECEDES:
         ^
lib/nl_langinfo.c:301:10: error: use of undeclared identifier 'N_CS_PRECEDES'
    case N_CS_PRECEDES:
         ^
lib/nl_langinfo.c:303:10: error: use of undeclared identifier 'P_SEP_BY_SPACE'
    case P_SEP_BY_SPACE:
         ^
lib/nl_langinfo.c:305:10: error: use of undeclared identifier 'N_SEP_BY_SPACE'
    case N_SEP_BY_SPACE:
         ^
lib/nl_langinfo.c:307:10: error: use of undeclared identifier 'P_SIGN_POSN'
    case P_SIGN_POSN:
         ^
lib/nl_langinfo.c:309:10: error: use of undeclared identifier 'N_SIGN_POSN'
    case N_SIGN_POSN:
         ^
16 errors generated.
make[4]: *** [Makefile:9354: lib/nl_langinfo.o] Error 1
make[4]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
make[3]: *** [Makefile:11767: all-recursive] Error 1
make[3]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
make[2]: *** [Makefile:6306: all] Error 2
make[2]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools/coreutils'
make[1]: *** [Makefile:90: make] Error 2
make[1]: Leaving directory '/data/data/com.termux/files/home/dgsh_gm/dgsh/unix-tools'
make: *** [Makefile:81: unix-tools] Error 2
```

nl_langinfo related error

--- a/packages/dgsh/build.sh
+++ b/packages/dgsh/build.sh
@@ -1,0 +1,22 @@
+TERMUX_PKG_HOMEPAGE=https://www.spinellis.gr/sw/dgsh/
+TERMUX_PKG_DESCRIPTION="The directed graph shell"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=1.0.0
+TERMUX_PKG_SRCURL=https://github.com/dspinellis/dgsh/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=22a7f2794e1287a46b03ce38c27a1d9349d1c66535c30e065c8783626555c76c
+TERMUX_PKG_BUILD_DEPENDS="check"
+TERMUX_PKG_BUILD_IN_SRC=true
+
+termux_step_pre_configure() {
+	TERMUX_PKG_SRCDIR=$TERMUX_PKG_SRCDIR/core-tools
+	TERMUX_PKG_BUILDDIR=$TERMUX_PKG_SRCDIR
+	cd $TERMUX_PKG_BUILDDIR
+
+	sed -i -e 's/#.*$//g' src/dgsh-elf.s
+	cp $TERMUX_PKG_BUILDER_DIR/s_cpow.c src/
+
+	touch ../.config
+	mkdir -p m4
+	autoreconf -fi
+}

--- a/packages/dgsh/core-tools-src-dgsh-tee.c.patch
+++ b/packages/dgsh/core-tools-src-dgsh-tee.c.patch
@@ -1,0 +1,11 @@
+--- a/core-tools/src/dgsh-tee.c
++++ b/core-tools/src/dgsh-tee.c
+@@ -439,7 +439,7 @@
+ #ifdef FALLOC_FL_PUNCH_HOLE
+ 	static bool warned = false;
+ 
+-	if (fallocate(tmp_fd, FALLOC_FL_PUNCH_HOLE, pool * buffer_size, buffer_size) < 0 &&
++	if (fallocate(bp->page_file_fd, FALLOC_FL_PUNCH_HOLE, pool * buffer_size, buffer_size) < 0 &&
+ 	    !warned) {
+ 		warn("Failed to free temporary buffer space");
+ 		warned = true;

--- a/packages/dgsh/core-tools-src-dgsh-w.c.patch
+++ b/packages/dgsh/core-tools-src-dgsh-w.c.patch
@@ -1,0 +1,11 @@
+--- a/core-tools/src/dgsh-w.c
++++ b/core-tools/src/dgsh-w.c
+@@ -10,6 +10,8 @@
+ #include "dgsh.h"
+ #include "dgsh-debug.h"
+ 
++#include "s_cpow.c"
++
+ void
+ read_number(int fd, long double *x, long double complex *xc)
+ {

--- a/packages/dgsh/s_cpow.c
+++ b/packages/dgsh/s_cpow.c
@@ -1,0 +1,82 @@
+/*-
+ * Copyright (c) 2008 Stephen L. Moshier <steve@moshier.net>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*							cpow
+ *
+ *	Complex power function
+ *
+ *
+ *
+ * SYNOPSIS:
+ *
+ * double complex cpow();
+ * double complex a, z, w;
+ *
+ * w = cpow (a, z);
+ *
+ *
+ *
+ * DESCRIPTION:
+ *
+ * Raises complex A to the complex Zth power.
+ * Definition is per AMS55 # 4.2.8,
+ * analytically equivalent to cpow(a,z) = cexp(z clog(a)).
+ *
+ * ACCURACY:
+ *
+ *                      Relative error:
+ * arithmetic   domain     # trials      peak         rms
+ *    IEEE      -10,+10     30000       9.4e-15     1.5e-15
+ *
+ */
+
+#include <sys/cdefs.h>
+__FBSDID("$FreeBSD$");
+
+#include <complex.h>
+#include <float.h>
+#include <math.h>
+
+#if defined __ANDROID__ && __ANDROID_API__ < 26
+
+#pragma GCC visibility push(hidden)
+
+double complex
+cpow(double complex a, double complex z)
+{
+	double complex w;
+	double x, y, r, theta, absa, arga;
+
+	x = creal (z);
+	y = cimag (z);
+	absa = cabs (a);
+	if (absa == 0.0) {
+		return (CMPLX(0.0, 0.0));
+	}
+	arga = carg (a);
+	r = pow (absa, x);
+	theta = x * arga;
+	if (y != 0.0) {
+		r = r * exp (-y * arga);
+		theta = theta + y * log (absa);
+	}
+	w = CMPLX(r * cos (theta),  r * sin (theta));
+	return (w);
+}
+
+#pragma GCC visibility pop
+
+#endif

